pkg/capi2mapi/aws.go

```diff
@@ -0,0 +1,432 @@
+		return mapiv1.MachineSet{}, nil, errCAPIMachineAWSMachineTemplateAWSClusterCannotBeNil
+	}
+
+	var (
+		errors   []error
+		warnings []string
+	)
+
+	mapaSpec, warn, err := FromMachineAndAWSMachineTemplateAndAWSCluster(
+		&capiv1.Machine{
+			Spec: m.MachineSet.Spec.Template.Spec,
+			ObjectMeta: metav1.ObjectMeta{
+				Annotations: m.MachineSet.ObjectMeta.Annotations,
+			},
+		}, m.Template, m.AWSCluster).ToProviderSpec()
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	warnings = append(warnings, warn...)
+
+	mapiMachineSet, warn, err := FromMachineSetToMachineSet(m.MachineSet)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	warnings = append(warnings, warn...)
+
+	awsRawExt, err := RawExtensionFromProviderSpec(mapaSpec)
+	if err != nil {
+		errors = append(errors, err)
+	}
+
+	mapiMachineSet.Spec.Template.Spec.ProviderSpec.Value = awsRawExt
+
+	if len(errors) > 0 {
+		return mapiv1.MachineSet{}, warnings, utilerrors.NewAggregate(errors)
+	}
+
+	return mapiMachineSet, warnings, nil
+}
+
+// Conversion helpers.
+
+// RawExtensionFromProviderSpec marshals the machine provider spec.
+func RawExtensionFromProviderSpec(spec *mapiv1.AWSMachineProviderConfig) (*runtime.RawExtension, error) {
+	if spec == nil {
+		return &runtime.RawExtension{}, nil
+	}
+
+	var (
+		rawBytes []byte
+		err      error
+	)
+
+	if rawBytes, err = json.Marshal(spec); err != nil {
+		return nil, fmt.Errorf("error marshalling providerSpec: %w", err)
+	}
+
+	return &runtime.RawExtension{
+		Raw: rawBytes,
+	}, nil
+}
+
+// AWSProviderSpecFromRawExtension unmarshals a raw extension into an AWSMachineProviderSpec type.
+func AWSProviderSpecFromRawExtension(rawExtension *runtime.RawExtension) (mapiv1.AWSMachineProviderConfig, error) {
+	if rawExtension == nil {
+		return mapiv1.AWSMachineProviderConfig{}, nil
+	}
+
+	spec := mapiv1.AWSMachineProviderConfig{}
+	if err := yaml.Unmarshal(rawExtension.Raw, &spec); err != nil {
+		return mapiv1.AWSMachineProviderConfig{}, fmt.Errorf("error unmarshalling providerSpec: %w", err)
+	}
+
+	return spec, nil
+}
+
+func convertAWSMetadataOptionsToMAPI(capiMetadataOpts *capav1.InstanceMetadataOptions) (mapiv1.MetadataServiceOptions, []string, error) {
+	var (
+		errors   []error
+		warnings []string
+	)
+
+	if capiMetadataOpts == nil {
+		return mapiv1.MetadataServiceOptions{}, nil, nil
+	}
+
+	var auth mapiv1.MetadataServiceAuthentication
+
+	switch capiMetadataOpts.HTTPTokens {
+	case "":
+		//
+	case capav1.HTTPTokensStateOptional:
+		auth = mapiv1.MetadataServiceAuthenticationOptional
+	case capav1.HTTPTokensStateRequired:
+		auth = mapiv1.MetadataServiceAuthenticationRequired
+	default:
+		errors = append(errors, fmt.Errorf("%w: %q", errUnsupportedHTTPTokensState, capiMetadataOpts.HTTPTokens))
+	}
+
+	// TODO: lossy: These fields are not present in MAPI, so they will be lost in CAPI -> MAPI conversion.
+	// opts.HTTPEndpoint
+	// opts.HTTPPutResponseHopLimit
+	// opts.InstanceMetadataTags
```

Comment:
If any of these are present, they need to error out as the feature is not support in MAPA

Reply:
So these fields are required and always present, so we need to work out a suitable handling of this. Eg if the values are the defaults, then we should not error, if they are non default, we should error.

Should investigate how MAPI handles this, perhaps we want to add these on the MAPI side to save conversion